packages/components/src/QueryModel/withQueryModels.tsx

```diff
@@ -0,0 +1,325 @@
+import React, { ComponentType, PureComponent } from 'react';
+import produce from 'immer';
+
+import { resolveErrorMessage, SchemaQuery } from '..';
+import { LoadingState, QueryConfig, QueryModel } from './QueryModel';
+import { DefaultQueryModelLoader, QueryModelLoader } from './QueryModelLoader';
+
+export interface Actions {
+    addModel: (queryConfig: QueryConfig, load?: boolean) => void;
+    loadModel: (id: string) => void;
+    loadAllModels: () => void;
+    loadRows: (id: string) => void;
+    loadNextPage: (id: string) => void;
+    loadPreviousPage: (id: string) => void;
+    loadFirstPage: (id: string) => void;
+    loadLastPage: (id: string) => void;
+    setOffset: (id: string, offset: number) => void;
+    setMaxRows: (id: string, maxRows: number) => void;
+    setView: (id: string, viewName: string) => void;
+    setSchemaQuery: (id: string, schemaQuery: SchemaQuery) => void;
+}
+
+export interface RequiresModelAndActions {
+    model: QueryModel;
+    actions: Actions;
+}
+
+export interface InjectedQueryModels {
+    queryModels: { [key: string]: QueryModel };
+    actions: Actions;
+}
+
+export type QueryConfigMap = { [id: string]: QueryConfig };
+export type QueryModelMap = { [id: string]: QueryModel };
+
+export interface MakeQueryModels {
+    autoLoad?: boolean;
+    modelLoader?: QueryModelLoader;
+    queryConfigs?: QueryConfigMap;
+}
+
+interface State {
+    queryModels: QueryModelMap;
+}
+
+export function withQueryModels<Props>(ComponentToWrap: ComponentType<Props & InjectedQueryModels>)
+    : ComponentType<Props & MakeQueryModels> {
+    class ComponentWithQueryModels extends PureComponent<Props & MakeQueryModels, State> {
+        actions: Actions;
+        static defaultProps;
+
+        constructor(props: Props & MakeQueryModels) {
+            super(props);
+            const { queryConfigs } = props;
+
+            const queryModels = Object.keys(props.queryConfigs).reduce((models, id) => {
+                const queryConfig = {id, ...queryConfigs[id]};
+                models[id] = new QueryModel(queryConfig);
+                return models;
+            }, {});
+
+            const initialState: State = {
+                queryModels,
+            };
+
+            this.state = produce(initialState, () => {});
```

Comment:
I'm confused by this pattern.  What does initialState do here?

Reply:
Initial state is the state we want to initialize this component with, in this case it's the set of `QueryModels` instantiated from the passed in `QueryConfig`. The reason the second arg is a no-op function is because you have to pass in a function as the second arg to make changes to your state. If you don't pass a second arg it is assumed you're first arg is a function, and produce will return a [curried producer](https://immerjs.github.io/immer/docs/curried-produce) which we don't want in this case. The no-op is because we don't need to modify the initialState object at all. I could simplify this line to something like:

```
this.state = produce({ queryModels }, () => {});
```

Reply:
I simplified the code a bit, and added a note to clarify what the reducer above is doing as well.

Reply:
Got it ok.  Just a quick glance looked weird but makes sense now. Thanks.